CppCoreGuidelines.md

```diff
@@ -19181,27 +19181,27 @@ A header should include all its dependencies. Be careful about using relative pa
 
 A test should verify that the header file itself compiles or that a cpp file which only includes the header file compiles.
 
-### <a name="Rs-incform"></a>SF.12: Prefer the angle bracket form of `#include` where you can and the quoted form everywhere else
+### <a name="Rs-incform"></a>SF.12: Prefer the quoted form of `#include` for files relative to the including file and the angle bracket form everywhere else
 
 ##### Reason
 
 The [standard](http://eel.is/c++draft/cpp.include) provides flexibility for compilers to implement
 the two forms of `#include` selected using the angle (`<>`) or quoted (`""`) syntax. Vendors take
 advantage of this and use different search algorithms and methods for specifying the include path.
 
-Nevertheless, the guidance is to use the angle form when possible. This supports the fact that the
-standard library headers must be included this way, is more likely to create portable code, and enables
-the quoted form for other uses. For example being clear about the locality of the header relative
-to files that includes it or in scenarios where the different search algorithm is required.
+Nevertheless, the guidance is to use the quoted form for including files that exist at a relative path to the file containing the `#include` statement and to use the angle bracket form everywhere else, where possible. This encourages being clear about the locality of the header relative to files that include it, or scenarios where the different search algorithm is required. It makes it easy to understand at a glance whether a header is being included from a local relative file versus a standard library header or a header from the alternate search path (e.g. a header from another project or a common set of includes).
 
 ##### Example
 
-    #include <string>       // From the standard library, required form
```

Comment:
If we want to avoid the confusion that arises from use of 'projects', we could instead talk in terms of header search paths. It's a term that fits reasonably well with common toolchains such as [MSVC](https://docs.microsoft.com/en-us/cpp/preprocessor/hash-include-directive-c-cpp?view=vs-2019) and [GCC](https://gcc.gnu.org/onlinedocs/gcc-4.1.2/cpp/Search-Path.html).

Also, saying where the _including_ file is located might make things clearer. So perhaps:

> In a source file, _/mygame/player.cpp_ which wishes to include
> * a header from the same directory _/mygame/shape.h_, and
> * a header from a physics library, _/physics/shape.h_
>
> and where _/physics_ is passed to the compiler as a **search path**, then use:
>
> ```c++
> // quotation marks to include relative header, /mygame/shape.h
> #include "shape.h"
> 
> // and angle brackets to include a header from the search path, /physics/shape.h
> #include <shape.h>
> ```

Reply:
That seems needlessly ambiguous. I was under the impression it was common practice (boost does it at least) to insert a redundant folder into the include folder to mitigate this so you add boost/include to your search path, the full path to a header is boost/include/boost/shape.h, and you include <boost/shape.h>. I guess the different include syntaxes help when neither you nor your library author did not follow this convention?

Reply:
>>     #include "shape.h" // intends to include <mygame/shape.h>
>>     #include <shape.h> // intends to include <physics/shape.h>
>
> That seems needlessly ambiguous.

+1. This is what I meant when I said elsethread in this PR, "The bad situation I want to avoid is that you have two different search paths simultaneously in the same compilation process, so that `#include "foo.h"` and `#include <foo.h>` both appear in the codebase, but have different effects." I would prefer for the coder to write `<mygame/shape.h>` or `<physics/shape.h>` depending on their intent, and then set _the_ include path in their build system so that those angle-bracket lookups worked.

> I guess the different include syntaxes help when neither you nor your library author did not follow this convention?

Yes. The different include syntaxes help when neither _you_ nor your library author followed the convention. But _you_ always have the option to follow the convention (at least, _your project_ always has the option). I admit I've worked on projects where e.g. `"assert.h"` did something different from `<assert.h>`. But I've also worked on projects where `<mygame/string.h>` did something different from `<string.h>`; I liked the latter projects better.

Reply:
After the previous edit, there wasn't much reliance on the term 'project'. I scrubbed it even more in this recent edit.

I did want to completely leave out the term 'library' in some examples, since it's a concrete example of when you would want to use the "alternate" search path. I also did not want to go too deeply into filesystem structure or compiler options, leaving "alternate search path" to imply "however you need to configure your compiler to find non-local includes" and "locally relative" to imply however the file system considers local relativity. trying to be a bit vague here on purpose so that we don't step into things like compiler-specific switches or system-specific concepts of locality (symlinks, hardlinks, network paths)

Reply:
@Quuxplusone having a subdirectory convention is orthogonal. Perhaps that would be a good rule for another day (or document).

The example is organized this way to imply that _physics_ is *not* something which the author of _player.cpp_ is in control of. Can you suggest some way to make this clearer? Bear in mind that we need to account for things like _/usr/include/shape.h_ so I'm not sure a subdirectory convention is an adequate solution to the problem we're trying to help avoid here.

Reply:
@apenn-msft yes, this looks good! The use of "e.g." in the paragraph above is enough to make it clear that `<>` might be applicable *before* we get as far as libraries.